### ISSUE TITLE
npm run <script> does not work when folder has %2F in its name.

### ISSUE THREAD

**Issue by NicolasPelletier** [npm/npm#12446]
# Context

I am trying to make 'npm run <script>' work where
1. The folder of my application has `%2F` in its name
2. The <script> path is not specified in the package.json
3. The <script> is in './node_modules/.bin/'

If condition (1) is false ( meaning that my folder name is something like /tmp/folder1 ) then the script runs.
If condition (1) is true ( meaning that my folder name is something like /tmp/fold%2Fer1 ) then the npm fails to find the script. It seems to not be able to identify that the script to run is available in `./node_modules/.bin/`.
# To reproduce

To reproduce the problem, clone something in a folder with `%2F` in its name. I made a simple repo for you. I clone it using destination folder `/tmp/GH%2FTest1`:

```
git clone git@github.com:NicolasPelletier/GHTest1.git /tmp/GH%2FTest1
cd /tmp/GH%2FTest1
npm install
npm run test
```
# The npm-debug log

I create a Gist to show my package.json and the debug log resulting from running `npm run test`: https://gist.github.com/NicolasPelletier/db97e1a4f25b2b3741d6224845cf51da
# Use Case

When creating branches in git, my team uses branches with `/` in them because it seems to make a difference to them when presented in SourceTree. When these branches are imported by Jenkins for CI testing, it attempts to create a folder on the machine with the code in it. To allow multiple users running tests at the same time, it creates a folder with the name of the branch. To limit collisions I think.  Jenkins cannot create a folder with the branch name unless it replaces the `/` by `%2F`.

So, in the next step, when Jenkins tries to run `npm run some_script`, it fails.

Note: I don't have control on the Jenkins machine. Another team is responsible for it so I don't know exactly how it is setup.
# Work Around

We can use complete paths in the package.json. This way, `npm run` does not have to try to resolve the path of the executable.

In the example below, replacing the simple `eslint .` by the complete path `./node_modules/eslint/bin/eslint.js .` solves the problem.

``` javascript
"scripts": {
   "test": "eslint .",  // This fails with ENOENT
   "eslint": "./node_modules/eslint/bin/eslint.js ." // This executes perfectly
}
```
# Conclusion

It would be nice if NPM would work the same whatever the folder name in which the project is deployed.

keywords: 'npm run', 'paths with unusual character', 'URI encoded forward slash'

**Comment by martianmartian** [comment]
Get me off the fucking email list

Sent from my iPhone

On Apr 23, 2016, at 9:42 AM, Nicolas Pelletier <notifications@github.com<mailto:notifications@github.com>> wrote:

Context

I am trying to make 'npm run <script>' work where

  1.  The folder of my application has %2F in its name
  2.  The <script> path is not specified in the package.json
  3.  The <script> is in './node_modules/.bin/'

If condition (1) is false ( meaning that my folder name is something like /tmp/folder1 ) then the script runs.
If condition (1) is true ( meaning that my folder name is something like /tmp/fold%2Fer1 ) then the npm fails to find the script. It seems to not be able to identify that the script to run is available in ./node_modules/.bin/.

To reproduce

To reproduce the problem, clone something in a folder with %2F in its name. I made a simple repo for you. I clone it using destination folder /tmp/GH%2FTest1:

git clone git@github.com:NicolasPelletier/GHTest1.git /tmp/GH%2FTest1
cd /tmp/GH%2FTest1
npm install
npm run test


The npm-debug log

I create a Gist to show my package.json and the debug log resulting from running npm run test: https://gist.github.com/NicolasPelletier/db97e1a4f25b2b3741d6224845cf51da

Use Case

When creating branches in git, my team uses branches with / in them because it seems to make a difference to them when presented in SourceTree. When these branches are imported by Jenkins for CI testing, it attempts to create a folder on the machine with the code in it. To allow multiple users running tests at the same time, it creates a folder with the name of the branch. To limit collisions I think. Jenkins cannot create a folder with the branch name unless it replaces the / by %2F.

So, in the next step, when Jenkins tries to run npm run some_script, it fails.

Note: I don't have control on the Jenkins machine. Another team is responsible for it so I don't know exactly how it is setup.

Work Around

We can use complete paths in the package.json. This way, npm run does not have to try to resolve the path of the executable.

In the example below, replacing the simple eslint . by the complete path ./node_modules/eslint/bin/eslint.js . solves the problem.

"scripts": {
   "test": "eslint .",  // This fails with ENOENT
   "eslint": "./node_modules/eslint/bin/eslint.js ." // This executes perfectly
}

Conclusion

It would be nice if NPM would work the same whatever the folder name in which the project is deployed.

keywords: 'npm run', 'paths with unusual character', 'URI encoded forward slash'

-

**Comment by KenanY** [comment]
@martianmartian I see that you are a [watcher of this repository](https://github.com/npm/npm/watchers). This means that you are automatically subscribed to all issues and pull requests, and you will receive GitHub notifications for all comments. These notifications become emails if you have such a setting enabled in your GitHub settings. If you would like to stop watching this repository, you can easily just unwatch it:

![2016-04-23-15-58-10-5094586](https://cloud.githubusercontent.com/assets/733364/14763726/37847b16-096c-11e6-87fb-fcca17a165f1.png)


or configure your GitHub settings to not send emails.

**Comment by STRML** [comment]
Can reproduce this - seeing CI breakage on e.g. feature branches with folder names like `feature/foo`.

**Comment by STRML** [comment]
This may actually be a core bug. Running the command from my terminal with the env.PATH output from `--verbose` works correctly.

**Comment by francisdb** [comment]
I can confirm this issue, automated github pull request builds fail on Jenkins because of the forward slash escape (`fix/parsetoint` pull request name)
we are using npm v2.15.1

**Comment by francisdb** [comment]
small remark, pointing to `./node_modules/.bin/binary` (symlink) also doesn't work

**Comment by ekmartin** [comment]
I can't seem to reproduce this using npm version 3.8.9 and node version 6.2.0. 
Both `npm run test` and `npm run lint` complete successfully.

**Comment by francisdb** [comment]
Just switched to the node_6 npm repo and we still have the issue:

```
sh: 1: jshint: not found
ignoring exit code 127
sh: 1: karma: not found

npm ERR! Linux 4.2.0-25-generic
npm ERR! argv "/usr/bin/nodejs" "/usr/bin/npm" "run" "ci"
npm ERR! node v6.2.2
npm ERR! npm  v3.9.5
```

our ci script in "package.json" looks like this 

```json
"scripts": {
    "ci": "npm install; npm run jshint; karma start test/karma.conf.js --no-auto-watch --single-run --reporters=dots,junit --browsers=Firefox,PhantomJS"
}
```

and we have all the needed (cli) dependencies in devDependencies and they work when not called from the package scripts

```
./node_modules/.bin/karma --help
Karma - Spectacular Test Runner for JavaScript.

Usage:
  /usr/bin/nodejs ./node_modules/.bin/karma <command>
...
```

**Comment by eppsilon** [comment]
We also experienced this issue on a Jenkins build system running Debian 8 with Node 6.2.2 and NPM 3.9.5.

Replacing the commands in the NPM scripts seemed to work at first, but that didn't help when a dependency (node-sass) needed to run something (`node scripts/install.js`) upon install.

However, I was able to find another workaround after seeing this [thread about "%" characters in $PATH](http://unix.stackexchange.com/questions/126955/percent-in-path-environment-variable). NPM runs scripts using `child_process`, which ultimately uses `/bin/sh -C` to run a command. Normally `/bin/sh` points at Ash or Dash, so we pointed it at Bash instead. That fixed the issue.

Here's the commands we used to change `/bin/sh`. We used them in a Dockerfile. Non-Debian systems probably have an equivalent command.

```
echo "dash dash/sh boolean false" | debconf-set-selections
DEBIAN_FRONTEND=noninteractive dpkg-reconfigure dash
```

**Comment by STRML** [comment]
Very interesting  - nice find @eppsilon!

**Comment by francisdb** [comment]
thanks @eppsilon that indeed seems to be a valid workaround

**Comment by NicolasPelletier** [comment]
At the time I opened the bug, I did not have this work around on Jenkins, and since this was an npm issue ( or so I thought ), I did think of mention it here later. However, since people are mentioning Jenkins on this thread now, here is what I finally did to make it work. I used a workspace:
```
/**
 * - This is a workaround for https://github.com/npm/npm/issues/12446.
 */
private def getWorkspace() {
    pwd().replace('%2F', '_')
}

ws(getWorkspace()) {
  checkout scm
  npm install
  npm run stuff
}
```

By creating a new workspace, the folder in which npm runs does not contain %2F anymore and it works.

This does not work for other %xx formats, but it solved my immediate problem so that's what I did. It's been working fine so far since my developers never use % in their folder names and it was Jenkins that introduces the %2F when my devs used `/` in their branch names when they submitted Pull Requests.
